segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/SystemJournal.java

```diff
@@ -1111,6 +1114,41 @@ private CompletableFuture<Void> applyChunkAddition(MetadataTransaction txn, Map<
                 }, executor);
     }
 
+    /**
+     * Apply chunk append.
+     */
+    private CompletableFuture<Void> applyAppend(MetadataTransaction txn, String segmentName, String chunkName, long offset, long length) {
+        Preconditions.checkState(null != chunkName && !chunkName.isEmpty(), "chunkName must not be null or empty");
+        Preconditions.checkState(offset >= 0, "offset must be non-negative");
+        Preconditions.checkState(length > 0, "length must be positive");
+
+        return validateSegment(txn, segmentName)
+                .thenComposeAsync(v -> txn.get(segmentName), executor)
+                .thenComposeAsync(m -> {
+                    val segmentMetadata = (SegmentMetadata) m;
+                    segmentMetadata.checkInvariants();
+                    // set length.
+                    return txn.get(segmentMetadata.getLastChunk())
+                            .thenAcceptAsync( metadata -> {
+                                Preconditions.checkState(metadata != null, "metadata must not be null");
+                                ChunkMetadata lastChunkMetadata = (ChunkMetadata) metadata;
+                                Preconditions.checkState(lastChunkMetadata.getName().equals(chunkName), "Invalid chunk name expected= %s actual=%s in segment=%s",
+                                        lastChunkMetadata.getName(), chunkName, segmentMetadata);
+                                Preconditions.checkState(lastChunkMetadata.getLength() == offset, "Invalid offset expected= %s actual=%s in segment=%s",
+                                        lastChunkMetadata.getLength(), offset, segmentMetadata);
+                                segmentMetadata.setLength(segmentMetadata.getLength() + length);
+                                lastChunkMetadata.setLength(lastChunkMetadata.getLength() + length);
+                                segmentMetadata.setAtomicWrites(true);
```

Comment:
does setAtomicWrites need to be set everytime we update the journal?

Reply:
Segment metadata for all segment is stored in metadata tables. Once marked in metadata we do not need to mark it again.
1. For all new segments we mark appropriately during create operation
2. For old segments we update that flag when we open that segment for the first time after new code and update the metadata 
3. For system segments, the system snapshot the metadata is in system journals. We recreate in-memory metadata for system segments by replaying journals during boot. This particular line is called on bootstrap and always mark system segments as atomic